Make use of CMakeDebHelper to improve CPack output (#2790)
mixxxdj/mixxx: CMakeLists.txt

endif()

# Packaging

> set(CPACK_PACKAGE_NAME "mixxx")

Reviewer: This defaults to the project name, so we could remove this as it's unneeded.

set(CPACK_STRIP_FILES ON)
set(CPACK_CREATE_DESKTOP_LINKS "mixxx")

> set(CPACK_DEBIAN_PACKAGE_MAINTAINER "${CPACK_PACKAGE_CONTACT}")

Reviewer: ping

set(CPACK_DEBIAN_PACKAGE_SECTION "sound")
set(CPACK_DEBIAN_PACKAGE_PRIORITY "optional")
set(CPACK_DEBIAN_PACKAGE_SUGGESTS "pdf-viewer")
set(CPACK_DEBIAN_PACKAGE_REPLACES "mixxx-data")
set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5sql5-sqlite, pulseaudio-utils")
set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)

> set(CPACK_DEBIAN_PACKAGE_VERSION "2.4.0~alpha~pre-ppa1~master~git7371~bionic")

Author: We need stick with the curious version string for our deb file to fit into the > check. 
This doe not make sense for our other builds target. 
Fixing this should be done in a separate PR. The big underlying issue is that the git version is calculated at configure time, but not at pack time. 
We need a solution that uses the version from the binary that is packed and not an outdated version from the configure time. 
Do you know how to achieve this?

> set(CPACK_DEBIAN_PACKAGE_VERSION "2.4.0~alpha~pre-ppa1~master~git7371~bionic")

Reviewer: `CPackConfig` runs at configure time. But I don't think we should overwrite this variable at all. What I want to build a package for debian?

Instead, we should invoke it like this: cpack -G -DCPACK_PACKAGE_VERSION=foo

> set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "${CPACK_PACKAGE_HOMEPAGE_URL}")

Author: It does not work here unfortunately.
So I set it explicit.

> set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "${CPACK_PACKAGE_HOMEPAGE_URL}")

Reviewer 2: Hard-coding distribution-specific build versions into the generic build file doesn't seem to be right.

> set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "${CPACK_PACKAGE_HOMEPAGE_URL}")

Author: Yes, I know. This is s place holder to generate the git version from cmake.
This is quite a bit work I want to avoid now.

set(CPACK_DEBIAN_PACKAGE_DESCRIPTION

> "Digital Disc Jockey Interface

Reviewer 3: ```suggestion
"Digital Disc Jockey Application
```
?

 Mixxx is a digital DJ system, where Wave, Ogg, FLAC and MP3 files can be
 mixed on a computer for use in live performances. Filters, crossfader, and

>  speed control are provided. Mixxx can sync the 2 streams automatically, using

Reviewer 3: ```suggestion
 speed control are provided. Mixxx can sync the 4 decks automatically, using
```

 an algorithm to detect the beat.

>  .

Reviewer 3: Does this `.` need to be here?

Author: Yes, that is a debian hack for an empty line.

>  Mixxx works with Jack or OSS, can be controlled from the GUI or from external

Reviewer 3: ```suggestion
 Mixxx works with ALSA or JACK, can be controlled from the GUI or from external
```

>  controllers, including MIDI devices and joysticks, and supports skins.")

Author: This also does not work by default.
I have tried to read the description file but that produced a broken deb.
Than I noticed that the debian maintainers rip our long description anyway. So I have decided to use the description of the debian maintainers directly. 
Hopefully this way it is back under our control.

>  controllers, including MIDI devices and joysticks, and supports skins.")

Reviewer: Can you update the description file and use `file(READ ...)` to read the file contents into a variable? That should work around the description file issue and keeps the description in a single place.

>  controllers, including MIDI devices and joysticks, and supports skins.")

Author: You mean the short description for all?

>  controllers, including MIDI devices and joysticks, and supports skins.")

Reviewer: > You mean the short description for all?

I don't understand your question, can you rephrase?

>  controllers, including MIDI devices and joysticks, and supports skins.")

Author: Do you mean we should discard our long description and use the shorted description from the Debian package for all our packages?

>  controllers, including MIDI devices and joysticks, and supports skins.")

Reviewer: Yes.

set(CPACK_WIX_UI_BANNER "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/banner.bmp")
set(CPACK_WIX_UI_DIALOG "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/dialog.bmp")

if(UNIX AND NOT APPLE)
  add_subdirectory(build/debian)
  add_subdirectory(build/rpm)

> endif()

Reviewer: The code from these `CMakeLists.txt` files should go into a `CPackConfigDEB.cmake` and `CPackConfigRPM.cmake` inside the `cmake` directory instead.

Then, you can include them from the `CPackConfig.cmake` file (not `CMakeLists.txt`) like this:

    if(CPACK_GENERATOR MATCHES "DEB")
        include(CPackConfigDEB)
    endif()

Author: I am not sure if this works, because CPackConfig.cmake is used at packaging time. Did you try it? 

But anyway, the current structure is done in a way that the project specific settings are in CMakeList.txt files in context at the debian and rpm folder while the common files that can be used in any project live in the cmake folder. I would like to keep this. 

What is the bad thing with add_subdirectory here?

# Packaging
set(CPACK_PACKAGE_VENDOR "Mixxx Project")
set(CPACK_PACKAGE_CONTACT "RJ Skerry-Ryan <rryan@mixxx.org>")

> set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Mixxx, powerfull features for all DJs")

Reviewer 3: ```suggestion
set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Mixxx DJ software")
```

Reviewer 3: No need for embellishing language here, just a plain description is more helpful IMO

set(CPACK_WIX_UI_BANNER "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/banner.bmp")
set(CPACK_WIX_UI_DIALOG "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/dialog.bmp")

include(CPack)

> if(APPLE AND MACOS_BUNDLE)

Reviewer 3: Why did you remove macOS bundle support?

Author: Ups, that was not intended, a rebasing issue. I will rebase again.
